File: examples/go/servers/advanced/eip2612-gas-sponsoring.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"time"

	ginfw "github.com/gin-gonic/gin"
	"github.com/joho/godotenv"
	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/extensions/eip2612gassponsor"
	x402http "github.com/x402-foundation/x402/go/http"
	ginmw "github.com/x402-foundation/x402/go/http/gin"
	evm "github.com/x402-foundation/x402/go/mechanisms/evm/exact/server"
)

const eip2612DefaultPort = "4021"

// This example demonstrates how to set up a server that supports the
// Permit2 payment flow with EIP-2612 gasless approval. When a client
// doesn't have an existing Permit2 approval, the facilitator will use
// the client's EIP-2612 permit signature to approve Permit2 on-chain
// as part of the settlement transaction.
func eip2612GasSponsoringExample() {
	_ = godotenv.Load()

	facilitatorURL := os.Getenv("FACILITATOR_URL")
	if facilitatorURL == "" {
		fmt.Println("FACILITATOR_URL is required")
		os.Exit(1)
	}

	evmPayeeAddress := os.Getenv("EVM_ADDRESS")
	if evmPayeeAddress == "" {
		fmt.Println("EVM_ADDRESS is required")
		os.Exit(1)
	}

	port := os.Getenv("PORT")
	if port == "" {
		port = eip2612DefaultPort
	}

	evmNetwork := x402.Network("eip155:84532")
	facilitatorClient := x402http.NewHTTPFacilitatorClient(facilitatorURL)

	// Build EIP-2612 gas sponsoring extension declaration
	eip2612Ext := eip2612gassponsor.DeclareEip2612GasSponsoringExtension()
	extensions := make(map[string]interface{})
	for k, v := range eip2612Ext {
		extensions[k] = v
	}

	// Define routes with Permit2 + EIP-2612 gas sponsoring
	routes := x402http.RouteConfig{
		// This endpoint uses Permit2 and advertises EIP-2612 gas sponsoring
		"GET /premium-data": {
			Accepts: x402http.PaymentOptions{
				{
					Scheme:  "exact",
					PayTo:   evmPayeeAddress,
					Network: evmNetwork,
					Price: map[string]interface{}{
						"amount": "1000",
						"asset":  "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
						"extra": map[string]interface{}{
							"assetTransferMethod": "permit2",
						},
					},
				},
			},
			Extensions: extensions,
		},
	}

	r := ginfw.Default()

	r.Use(ginmw.X402Payment(ginmw.Config{
		Routes:      routes,
		Facilitator: facilitatorClient,
		Schemes: []ginmw.SchemeConfig{
			{Network: evmNetwork, Server: evm.NewExactEvmScheme()},
		},
		SyncFacilitatorOnStart: true,
		Timeout:                30 * time.Second,
	}))

	r.GET("/premium-data", func(c *ginfw.Context) {
		c.JSON(http.StatusOK, ginfw.H{
			"data":      "premium market data",
			"timestamp": time.Now().Format(time.RFC3339),
		})
	})

	fmt.Printf("Server with EIP-2612 Gas Sponsoring listening at http://localhost:%s\n", port)
	if err := r.Run(":" + port); err != nil {
		fmt.Printf("Failed to start server: %v\n", err)
		os.Exit(1)
	}
}

File: examples/typescript/servers/advanced/eip2612-gas-sponsoring.ts
/**
 * Example: Server with EIP-2612 Gas Sponsoring Extension
 *
 * This example demonstrates how to set up a server that supports the
 * Permit2 payment flow with EIP-2612 gasless approval. When a client
 * doesn't have an existing Permit2 approval, the facilitator will use
 * the client's EIP-2612 permit signature to approve Permit2 on-chain
 * as part of the settlement transaction.
 *
 * Required environment variables:
 * - EVM_ADDRESS: The payee wallet address
 * - FACILITATOR_URL: URL of the facilitator service
 */

import { config } from "dotenv";
import express from "express";
import { paymentMiddleware, x402ResourceServer } from "@x402/express";
import { ExactEvmScheme } from "@x402/evm/exact/server";
import { HTTPFacilitatorClient } from "@x402/core/server";
import { declareDiscoveryExtension } from "@x402/extensions/bazaar";
import { declareEip2612GasSponsoringExtension } from "@x402/extensions";
config();

const evmAddress = process.env.EVM_ADDRESS as `0x${string}`;
if (!evmAddress) {
  console.error("Missing EVM_ADDRESS environment variable");
  process.exit(1);
}

const facilitatorUrl = process.env.FACILITATOR_URL;
if (!facilitatorUrl) {
  console.error("Missing FACILITATOR_URL environment variable");
  process.exit(1);
}
const facilitatorClient = new HTTPFacilitatorClient({ url: facilitatorUrl });

const app = express();

app.use(
  paymentMiddleware(
    {
      // This endpoint uses the Permit2 asset transfer method
      // and declares support for EIP-2612 gasless approval.
      // Clients that don't have an existing Permit2 approval can
      // use the EIP-2612 extension to get a gasless approval.
      "GET /premium-data": {
        accepts: {
          scheme: "exact",
          network: "eip155:84532",
          payTo: evmAddress,
          // Use pre-parsed price to force Permit2 flow
          price: {
            amount: "1000", // 0.001 USDC
            asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
            extra: {
              assetTransferMethod: "permit2",
            },
          },
        },
        description: "Premium market data with gasless Permit2 approval",
        mimeType: "application/json",
        extensions: {
          // Advertise Bazaar discovery (optional)
          ...declareDiscoveryExtension({
            output: {
              example: { data: "premium market data" },
            },
          }),
          // Advertise EIP-2612 gas sponsoring support
          // This tells clients that the facilitator will accept an
          // EIP-2612 permit signature for gasless Permit2 approval
          ...declareEip2612GasSponsoringExtension(),
        },
      },
    },
    new x402ResourceServer(facilitatorClient).register("eip155:84532", new ExactEvmScheme()),
  ),
);

app.get("/premium-data", (req, res) => {
  res.json({
    data: "premium market data",
    timestamp: new Date().toISOString(),
  });
});

app.listen(4021, () => {
  console.log("Server with EIP-2612 Gas Sponsoring listening at http://localhost:4021");
});
